File: prisioner/js/judge/Judge.js
/**
 * http://usejsdoc.org/
 */
var Decision = require('../prisoner/Decision.js').Decision
var DECISION_OPTIONS = require('../prisoner/Decision.js').DECISION_OPTIONS
var CONSTANTS = require('./Constants.js').CONSTANTS

var Judge = function() {
	/*
	 * This class made a confession for every round
	 * storing the decisions made for every player into a list. 
	 */	
	this._paco;
	this._anton;
	this._decisionsPaco = new Array();
	this._decisionsAnton = new Array();
	
	this.setPaco = function(paco){
		this._paco = paco;
	}
	
	this.setAnton = function(anton){
		this._anton = anton;
	}
	this.judgementRun = function() {
		/*
		 * Executes the procedure for every round
		 * 1- Clears the previous decisions lists if there was one.
		 * 2- Make both prisoners confess N times
		 * 3- Compute with the decisions lists the points of both prisoners.
		 */
		this._decisionsPaco.length = 0;
		this._decisionsAnton.length = 0;
		this.confess();
		var puntuations = this.computePoints();
		return puntuations;
	};
	
	this.confess = function() {
		/*
    	 * Confess just execute the decide method from both prisoners
    	 * and stores the result in a list.
    	 */
		var pacoDecision, antonDecision;
		for(var i = 0; i < CONSTANTS.NUMBER_CONFESSIONS; i++){
			pacoDecision = this._paco.decide(this._decisionsPaco);
			antonDecision = this._anton.decide(this._decisionsAnton);
			this._decisionsPaco.push(new Decision(pacoDecision, antonDecision));
			// Disguises Anton as Paco by reversing the order, so Anton he knows which decision has made.
			this._decisionsAnton.push(new Decision(antonDecision, pacoDecision));
		}
		
	}

	this.computePoints = function() {
		/*
    	 * Returns both points 
    	 * after consulting the list of the decisions after the confession.
    	 */
		var accPaco = 0;
		var accAnton = 0;
		// Only iterates through decisionsPaco
    	// Decisions are the same for both lists excepts order is reversed.
		this._decisionsPaco.forEach(decision => {
			var numericDecisionPaco =decision.getPacoDecision();
			var numericDecisionAnton =decision.getAntonDecision();
			// If both decisions were the same
			if(numericDecisionPaco == numericDecisionAnton){
				if(numericDecisionPaco==DECISION_OPTIONS.CONFESS || numericDecisionPaco==DECISION_OPTIONS.BRIBERY) {
					accPaco += CONSTANTS.UNLUCKY_BOTH_YEARS;
					accAnton += CONSTANTS.UNLUCKY_BOTH_YEARS;
				} else {
					accPaco += CONSTANTS.LUCKY_BOTH_YEARS;
					accAnton += CONSTANTS.LUCKY_BOTH_YEARS;
				}
			} else {
				// Decisions differs
				var luckyForPaco = this._isLuckyForPaco(numericDecisionPaco, numericDecisionAnton);
				if(luckyForPaco==true){
					accPaco += CONSTANTS.LUCKY_YEARS_CONFESSION;
					accAnton += CONSTANTS.BAD_YEARS_CONFESSION;
					
				} else {
					// Anton Was lucky this time.
					accAnton += CONSTANTS.LUCKY_YEARS_CONFESSION;
					accPaco += CONSTANTS.BAD_YEARS_CONFESSION;
				}

			}
		});
		
		return [accPaco, accAnton];
	}

	this._isLuckyForPaco = function(numericDecisionPaco, numericDecisionAnton) {
		var luckyForPaco = false;
		var confessAndNegation = ((numericDecisionPaco== DECISION_OPTIONS.CONFESS)&&(numericDecisionAnton==DECISION_OPTIONS.NEGATION));
		var briberyAndConfess = ((numericDecisionPaco== DECISION_OPTIONS.BRIBERY)&&(numericDecisionAnton==DECISION_OPTIONS.CONFESS));
		var negationAndBribery = ((numericDecisionPaco== DECISION_OPTIONS.NEGATION)&&(numericDecisionAnton==DECISION_OPTIONS.BRIBERY));
		if(confessAndNegation|| briberyAndConfess || negationAndBribery) luckyForPaco=true;
		return luckyForPaco;
	}
};	
module.exports.Judge = Judge;

File: prisioner/java/src/main/java/com/everis/competition/judge/Judge.java
package com.everis.competition.judge;

import java.util.ArrayList;

import com.everis.competition.prisoner.Decision;
import com.everis.competition.prisoner.IPrisoner;

public class Judge {
	/*
	 * This class made a confession for every round
	 * storing the decisions made for every player into a list. 
	 */	
	IPrisoner paco;
	IPrisoner anton;
	ArrayList<Decision> decisionsPaco;
	ArrayList<Decision> decisionsAnton;
	public Judge() {
		decisionsPaco = new ArrayList<Decision>();
		decisionsAnton = new ArrayList<Decision>();
	}
	
	public int[] judgmentRun() {
		/*
		 * Executes the procedure for every round
		 * 1- Clears the previous decisions lists if there was one.
		 * 2- Make both prisoners confess N times
		 * 3- Compute with the decisions lists the points of both prisoners.
		 */
		decisionsPaco.clear();
		decisionsAnton.clear();
		confess();
		int[] puntuations = computePoints(); 
        return puntuations;
	}
	public void setPaco(IPrisoner paco) {
		this.paco = paco;
	}
	
	public void setAnton(IPrisoner anton) {
		this.anton = anton;
	}

    private void confess() {
    	/*
    	 * Confess just execute the decide method from both prisoners
    	 * and stores the result in a list.
    	 */
        int pacoDecision, antonDecision;
        for (int i = 0; i < Constants.NUMBER_CONFESSIONS; i++) {
            pacoDecision = paco.decide(decisionsPaco);
            antonDecision = anton.decide(decisionsAnton);
            decisionsPaco.add(new Decision(pacoDecision, antonDecision));
            // Disguises Anton as Paco by reversing the order, so he knows which decision has made
            decisionsAnton.add(new Decision(antonDecision, pacoDecision));
        }
        
    }

    private int[] computePoints() {
    	/*
    	 * Returns both points 
    	 * after consulting the list of the decisions after the confession.
    	 */
    	int accPaco = 0;
    	int accAnton = 0;
    	// Only iterates through decisionsPaco
    	// Decisions are the same for both lists excepts order is reversed.
    	for(Decision decision: decisionsPaco) {
    		int decisionPaco = decision.getPacoDecision();
    		int decisionAnton = decision.getAntonDecision();
    		// If both decisions were the same
    		if(decisionPaco == decisionAnton) {
    			if(decisionPaco==Decision.CONFESS  || decisionPaco==Decision.BRIBERY) {
    				accPaco += Constants.UNLUCKY_BOTH_YEARS;
    				accAnton += Constants.UNLUCKY_BOTH_YEARS;
    			} else {
    				// Both decide to negate and dont confess
    				accPaco += Constants.LUCKY_BOTH_YEARS;
    				accAnton += Constants.LUCKY_BOTH_YEARS;
    			}
    			
    		} else {
    			// Decisions differs
    			boolean luckyForPaco = isLuckyForPaco(decisionPaco, decisionAnton);
    			if(luckyForPaco) {
    				accPaco += Constants.LUCKY_YEARS_CONFESSION;
    				accAnton += Constants.BAD_YEARS_CONFESSION;
    			} else {
    				// Anton was lucky this time
    				accAnton += Constants.LUCKY_YEARS_CONFESSION;
    				accPaco += Constants.BAD_YEARS_CONFESSION;
    			}
    			
    		}
 		
    	}
    	
        return new int[] {accPaco, accAnton};
    }
    
    private boolean isLuckyForPaco(int decisionPaco, int decisionAnton) {
    	boolean luckyPaco = false;
    	boolean confessAndNegation = ((decisionPaco==Decision.CONFESS)&&(decisionAnton==Decision.NEGATION)); 
    	boolean briberyAndConfess =((decisionPaco==Decision.BRIBERY)&&(decisionAnton==Decision.CONFESS)); 
    	boolean negationAndBribery = ((decisionPaco==Decision.NEGATION)&&(decisionAnton==Decision.BRIBERY));	
    	if(confessAndNegation || briberyAndConfess || negationAndBribery) luckyPaco = true;
    	return luckyPaco;
    }

}
